### DISCUSSION TITLE
Matter RGBW device, LoadAvg with PWM

### DISCUSSION THREAD

**Discussion by smhc** [arendst/Tasmota#20104] [Q&A]
I have a run of the mill RGBW Tuya light strip device modified with a ESP32-C3-12F that I am trying to control via Tasmota+Matter. I have configured it with 1 button and 4 PWM gpios and for the most part it works fine.

However I have two issues I am seeking some clarification on;

1. The LoadAvg jumps to 99 when it is "on". I understand Tasmota may be doing some processing to handle PWM however I assumed this would be done via hardware support instead, i.e adding very little load. Have I potentially misconfigured something that is causing it to use CPU instead?

2. I cannot seem to easily access the "W" of the RGBW via Google home. If I say "ok google change colour to bright white" it turns on the white light (and only the white light) with:

```
{"MtrReceived":{"1":{"Ep":1,"Hue":0,"Sat":0}}}
```

And the same is done if I click on just the right "cool white" section on the palette via the google home app, although it takes a few tries to find the exact spot. However when trying to set up an automation you can only choose from a fixed set of colour options, none of which will trigger the true white. The cool white option here appears to trigger the below, which also turns on the green of the RGB leds:

```
{"MtrReceived":{"1":{"Ep":1,"Hue":42,"Sat":41}}}
```

There is no "bright white" option which turns on just the white leds. I have had to work around this with a rule:

```
Rule1 on mtrreceived#1#Sat<50 do HsbColor2 0 endon on mtrreceived#1#Hue=42 do HsbColor2 0 endon
```

An alternative is to have a custom action which literally says "change colour to bright white". Both of which works ok. However am I missing something? Is it simply a case that Tasmota has only implemented support for RGB lights and not RGBW lights? Would the google home app show a different interface if it was detected as a RGBW device instead of RGB?

Some of my related settings are below, I've tried changing various settings to no avail:

```
stat/tasmota_58E99C/RESULT = {"WhiteBlend":"ON"}
stat/tasmota_58E99C/RESULT = {"RGBWWTable":"255,255,255,255,255"}
stat/tasmota_58E99C/RESULT = {"SetOption37":128}
stat/tasmota_58E99C/RESULT = {"LedTable":"ON"}
stat/tasmota_58E99C/RESULT = {"CTRange":"153,500"}
stat/tasmota_58E99C/RESULT = {"DimmerRange":{"Min":0,"Max":100}}
stat/tasmota_58E99C/RESULT = {"LedPwmMode1":"OFF"}
stat/tasmota_58E99C/RESULT = {"MultiPWM":"OFF"}
stat/tasmota_58E99C/RESULT = {"Palette":"[]"}
stat/tasmota_58E99C/RESULT = {"PWMCT":"OFF"}
stat/tasmota_58E99C/RESULT = {"VirtualCT":"You need to enable `SetOption106 1`"}
stat/tasmota_58E99C/RESULT = {"SetOption106":"OFF"}
stat/tasmota_58E99C/RESULT = {"SetOption107":"OFF"}
```

**Comment by smhc** [comment]
I have discovered that using "SetOption15 0" and controlling the pwms via "PWMx yyy" does not increase the LoadAvg yet still controls the LEDs in the same way.

Is there any reason why "SetOption15 1, Dimmerx y" uses substantially more CPU than "SetionOption15 0, PWMx y"? Is there anyway to have the CPU benefits of "SetOption15 0" with regular light control?

**Comment by s-hadinger** [comment]
There is currently no support for RGBW nor RGBCW in Tasmota Matter. I suggest to use white blended mode, see here : https://tasmota.github.io/docs/Lights/#white-blend-mode

With white blend mode, Matter sees a RGB and W channel is automatically adjusted. 

For cpu load, it may be due to `fade` being enabled which reduces the `sleep` duration. Can you check the value or `Sleep`?

**Comment by smhc** [comment]
Yep, I am using white blend mode and it mostly works. The Sleep setting is:

{"Sleep":{"200":{"Active":"10"}}}

If I say "ok google change the light colour to white" it works fine. The issue is that writing an "automation" in google home you can only pick from a minimal set of predefined colours like "cool white", "red", "warm white", "ivory", "snow" etc.. none of which gives you true white. As mentioned, if you write a custom action that actually asks for "white" it works ok. I am guessing this is because there is no support for RGBW as yet - that is fine. I am just confirming this.

I thought there might be a way to treat saturation levels below a certain level as being true white through the "RGBWWTable" or something similar.

I am using "Fade", but I have also tried having it turned off on a different board. The repro steps are quite simple:

1. Configure a pin as PWM
2. cmd: Power1 1
3. Wait 20 seconds...
4. cmd: Status 11
```
stat/tasmota_EC01C4/STATUS11 = {"StatusSTS":{"Time":"2023-11-26T21:18:34","Uptime":"0T00:02:48","UptimeSec":168,"Heap":151,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99
```
5. cmd: SetOption15 0
6. cmd: PWM1 1000
7. Wait 20 seconds...
8. cmd: Status 11
```
tat/tasmota_EC01C4/STATUS11 = {"StatusSTS":{"Time":"2023-11-26T21:20:06","Uptime":"0T00:00:31","UptimeSec":31,"Heap":151,"SleepMode":"Dynamic","Sleep":220,"LoadAvg":6,
```


I will raise an issue for this, as I am reasonably sure it is a bug based on https://tasmota.github.io/docs/Lights/#pwm6 which suggests hardware PWM should be used.

**Reply by s-hadinger** [reply]
I will try to reproduce, but there is no software PMW on esp32, only hardware

**Reply by smhc** [reply]
Yes, my issue https://github.com/arendst/Tasmota/issues/20109 got rejected with a similar response.

I saw a similar issue re: https://github.com/arendst/Tasmota/issues/13280 for esp8266. It also suggests LoadAvg is not CPU percentage. I still think a 99 LoadAvg is suggestive of high CPU.

Is there any way to check if you are using hardware vs software PWM?

**Reply by s-hadinger** [reply]
As mentioned, there is just no software PWM implementation on ESP32. The code does not exist.

**Comment by s-hadinger** [comment]
I cannot reproduce, I get a loadavg of `19` when PWM is off and `37` when it's on, only because `Sleep` goes from 50 to 25. I don't see any `99` value. Using multiple PWM doesn't change the value either.

**Comment by smhc** [comment]
Ok, I assume it is due to the different hardware as I did an erase/flash and still get the "issue". My HW is identified as "ESP32-D0WD-V3 v3.0", and its a "ESP-32S". I experienced the same thing on a ESP32-C3-12F.

The Sleep goes to 10 and LoadAvg to 99, after about 5-10 seconds or so. Regardless of the setting of SetOption60. I will try another ESP32 I have lying around.

From my other issue I get the impression "LoadAvg" doesn't matter.. but all the guides about conserving power suggest getting "Sleep" high and "LoadAvg" low.. but both shoot in the opposite direction when PWM is on.

**Reply by sfromis** [reply]
While `Sleep` length is indeed related to Tasmota CPU activity and a bit of power usage, `LoadAvg` is only an indicator, all other things equal, and with the value becoming **way** over 100 in some circumstances, it is just not a CPU load percentage. If you lower `Sleep`, values in `LoadAvg` will typically increase, unless other circumstances (like PWM) already make the number unusable.

**Comment by smhc** [comment]
However I am not lowering Sleep, Tasmota is simply by turning the light on. 
https://github.com/arendst/Tasmota/blob/fb51d9df8a141ce93fe1ce945bd0f681ac02686f/tasmota/my_user_config.h#L275

https://github.com/arendst/Tasmota/blob/fb51d9df8a141ce93fe1ce945bd0f681ac02686f/tasmota/tasmota_xdrv_driver/xdrv_04_light.ino#L1785

I assume lowering sleep to 10 to "avoid flicker" is only needed for fade and software based PWM. I'm not using fade or any kind of animation. Is it still needed for hardware based PWM? Does it have a negative impact to reduce sleep like this, potentially unnecessarily?

**Reply by s-hadinger** [reply]
`Sleep 10` is only needed to have a smoother Fade. It has no impact on flicker.

**Reply by sfromis** [reply]
You could change the default for `PWM_MAX_SLEEP` and check if you can measure a bit of difference in energy consumption.

**Reply by smhc** [reply]
> Sleep 10 is only needed to have a smoother Fade. It has no impact on flicker.

Unfortunately Tasmota sets it to 10 even if fade is disabled. And if fade is enabled, it keeps it at 10 even after the fade is finished. It's not possible to adjust this value without modifying Tasmota source. Perhaps the code should be:
```
if (Light.power && Light.fade_running) {
```

(although no doubt it also has to take into account various edge cases, light cycling, schemes etc. But for the general "on without fading/cycling" I would think a long sleep is fine).

> You could change the default for PWM_MAX_SLEEP and check if you can measure a bit of difference in energy consumption.

I assume there's some difference, otherwise there'd be no point in implementing anything other than sleep 10. (long term I do actually want the fade effect, so having it 10 during the fade then restored would be preferable to just changing the max). It's not just power consumption, but the warmth and life of the chip. As PWM controlling lights is pretty common, I would have thought getting it to "idle" as much as possible in this very common scenario would be desirable.
